File: ocaml/src/p0056.ml
(* Project Euler: Problem 56 *)

open Core

let find_max_digital_sum answer a b_max =
  let rec aux acc pow_ab =
    let pow_ab_lst = Euler.Util.z_digits pow_ab in
    if List.length pow_ab_lst * 9 < acc || Z.(equal pow_ab one)
    then acc
    else aux (Int.max (List.reduce_exn ~f:( + ) pow_ab_lst) acc) Z.(pow_ab / a)
  in
  aux answer (Z.pow a b_max)
;;

let compute upper =
  let a_lst =
    List.(
      range upper 2 ~stride:(-1) ~stop:`inclusive
      |> filter ~f:(fun n -> n mod 10 <> 0)
      |> map ~f:Z.of_int)
  in
  let rec aux ans = function
    | [] -> ans
    | x :: xs -> aux (find_max_digital_sum ans x upper) xs
  in
  aux 0 a_lst
;;

let solve () = compute 99 |> Int.to_string

(* Test *)

let%test_unit "Powerful Digit Sum" = [%test_eq: int] (compute 99) 972

File: rust/src/bin/p0056.rs
// Project Euler: Problem 56

euler::run_solver!(56);

fn solve() -> String {
    compute().to_string()
}

fn compute() -> usize {
    use num_bigint::BigUint;
    use std::cmp;

    let mut ans: usize = 0;

    for a in (1_u32..100).rev() {
        // assume that x = 10 * n
        // x^y = (10 * n)^y = 10^y * n^y, so sum(digits(x^y)) = sum(digits(n^y))
        // we can skip to check multiples of ten in this problem.
        if a % 10 == 0 {
            continue;
        }

        let a_big = BigUint::from(a);
        for b in (1..100).rev() {
            let p = a_big.pow(b);
            if BigUint::to_str_radix(&p, 10).len() * 9 < ans {
                break;
            }
            ans = cmp::max(ans, p.to_radix_le(10).into_iter().map(usize::from).sum());
        }
    }

    ans
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0056() {
        assert_eq!(compute(), 972);
    }
}
